site/app/models/gradeable/AutogradingConfig.php

```diff
@@ -0,0 +1,248 @@
+        $details['textboxes'] = parent::parseObject($this->textboxes);
+
+        return $details;
+    }
+
+    /**
+     * Gets the test cases for this configuration
+     * @return AutogradingTestcase[]
+     */
+    public function getTestcases() {
+        return $this->testcases;
+    }
+
+    /**
+     * Gets the text boxes for this configuration
+     * @return array
+     */
+    public function getTextboxes() {
+        return $this->getTextboxes();
+    }
+
+    /**
+     * Gets whether this config has an early submission incentive
+     * @return bool
+     */
+    public function hasEarlySubmissionIncentive() {
+        return $this->early_submission_incentive;
+    }
+
+    /* Disabled setters */
+
+    /** @internal */
+    public function setMaxSubmissionSize() {
+        throw new \BadFunctionCallException('Setters disabled for AutogradingConfig');
+    }
+
+    /** @internal */
+    public function setMaxSubmissions() {
+        throw new \BadFunctionCallException('Setters disabled for AutogradingConfig');
+    }
+
+    /** @internal */
+    public function setRequiredCapabilities() {
+        throw new \BadFunctionCallException('Setters disabled for AutogradingConfig');
+    }
+
+    /** @internal */
+    public function setMaxPossibleGradingTime() {
+        throw new \BadFunctionCallException('Setters disabled for AutogradingConfig');
+    }
+
+    /** @internal */
+    public function setPartNames() {
+        throw new \BadFunctionCallException('Setters disabled for AutogradingConfig');
+    }
+
+    /** @internal */
+    public function setEarlySubmissionMessage() {
+        throw new \BadFunctionCallException('Setters disabled for AutogradingConfig');
+    }
+
+    /** @internal */
+    public function setEarlySubmissionMinimumDaysEarly() {
+        throw new \BadFunctionCallException('Setters disabled for AutogradingConfig');
+    }
+
+    /** @internal */
+    public function setEarlySubmissionMinimumPoints() {
+        throw new \BadFunctionCallException('Setters disabled for AutogradingConfig');
+    }
+
+    /** @internal */
+    public function setEarlySubmissionTestCases() {
+        throw new \BadFunctionCallException('Setters disabled for AutogradingConfig');
+    }
+
+    /** @internal */
+    public function setTotalNonHiddenNonExtraCredit() {
+        throw new \BadFunctionCallException('Setters disabled for AutogradingConfig');
+    }
+
+    /** @internal */
+    public function setTotalNonHiddenExtraCredit() {
+        throw new \BadFunctionCallException('Setters disabled for AutogradingConfig');
+    }
+
+    /** @internal */
+    public function setTotalHiddenNonExtraCredit() {
+        throw new \BadFunctionCallException('Setters disabled for AutogradingConfig');
+    }
+
+    /** @internal */
+    public function setTotalHiddenExtraCredit() {
+        throw new \BadFunctionCallException('Setters disabled for AutogradingConfig');
+    }
```

Comment:
why not just set these to private and have an empty body?

Reply:
Stick:
```
if (method_exists($this, $name) && !is_callable($this, $name)) {
    throw new \BadFunctionCallException("{$name} is disabled for {$class_name}");
}
```
or something right after the `$class_name = get_class($this);` and then you can just mark all of these functions as private.

The next step I guess would be looking into metaprogramming the `@method` docs from the top of a class into what the allowed callable functions would be for a given class.

Reply:
I'll make this an issue and can add it in a dedicated PR since the latter option (which I think is a lot more elegant) means we'll need to verify all of the `@method` php docs for the model classes.